bpo-38741: Definition of multiple ']' in header configparser

jsnklln opened the pull request:
I've changed up the regex for finding section headers.
I don't know if this covers every case but it covers
all the cases I could come up with.  The docs don't really
call out if ] should be allowed in section headers but
I can't think of any reason it shouldn't be.

This change keeps the existing failure if you try and
define a section with no name, like [].

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-38741](https://bugs.python.org/issue38741) -->
https://bugs.python.org/issue38741
<!-- /issue-number -->

Commit:
bpo-38741: Definition of multiple ']' in header configparser

I've changed up the regex for finding section headers.
I don't know if this covers every case but it covers
all the cases I could come up with.  The docs don't really
call out if ] should be allowed in section headers but
I can't think of any reason it shouldn't be.

This change keeps the existing failure if you try and
define a section with no name, like [].

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

brandtbucher reviewed (changes requested):
The patch itself looks good, but I don't know enough about this module to decide if this behavior is technically correct or not. @ambv, thoughts?

Also, this probably needs a test or two. Can you include them in `test_configparser.py`?

Commit:
Added tests to ensure sections with ] in them are
read correctly.  Included a check for key=value because
the original issue mentioned that it might not be working.

Commit:
Merge branch 'configparser_handle_brace_in_header' of https://github.com/jsnklln/cpython into configparser_handle_brace_in_header

brandtbucher reviewed (approved):
Thanks @jsnklln! I recommended that #17086 be closed in favor of this, since it contains tests and handles the empty case.
